flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala

```diff
@@ -1943,6 +1943,147 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
       "false")
   }
 
+  @Test
+  def testTimestampDiff(): Unit = {
+    testTableApi(
+      "2018-07-05 11:11:11".toTimestamp - "2018-07-03 11:11:11".toTimestamp,
+      "'2018-07-05 11:11:11'.toTimestamp - '2018-07-03 11:11:11'.toTimestamp",
+      "172800000"
+    )
+
+    testTableApi(
+      "2018-07-05 11:11:11".toTimestamp - "2018-07-03".toDate,
+      "'2018-07-05 11:11:11'.toTimestamp - '2018-07-03'.toDate",
+      "213071000"
+    )
+
+    testTableApi(
+      "2018-07-03".toDate - "2018-07-05 11:11:11".toTimestamp,
+      "'2018-07-03'.toDate - '2018-07-05 11:11:11'.toTimestamp",
+      "-213071000"
+    )
+
+    testTableApi(
+      "2018-07-05".toDate - "2018-07-03".toDate,
+      "'2018-07-05'.toDate - '2018-07-03'.toDate",
+      "172800000"
+    )
+
+    testAllApis(
+      timestampDiff(
+        TimeIntervalUnit.DAY,
+        "2018-07-03 11:11:11".toTimestamp,
+        "2018-07-05 11:11:11".toTimestamp),
+      "timestampDiff(DAY, '2018-07-03 11:11:11'.toTimestamp," +
+        "'2018-07-05 11:11:11'.toTimestamp)",
+      "TIMESTAMPDIFF(DAY, " +
+        "TIMESTAMP '2018-07-03 11:11:11', TIMESTAMP '2018-07-05 11:11:11')",
+      "2")
+
+    testAllApis(
+      timestampDiff(TimeIntervalUnit.DAY, "2016-06-15".toDate, "2016-06-16 11:11:11".toTimestamp),
+      "timestampDiff(DAY, '2016-06-15'.toDate, '2016-06-16 11:11:11'.toTimestamp)",
+      "TIMESTAMPDIFF(DAY, DATE '2016-06-15', TIMESTAMP '2016-06-16 11:11:11')",
+      "1")
+
+    testAllApis(
+      timestampDiff(TimeIntervalUnit.DAY, "2016-06-15 11:00:00".toTimestamp, "2016-06-19".toDate),
+      "timestampDiff(DAY, '2016-06-15 11:00:00'.toTimestamp, '2016-06-19'.toDate)",
+      "TIMESTAMPDIFF(DAY, TIMESTAMP '2016-06-15 11:00:00', DATE '2016-06-19')",
+      "3")
+
+    testAllApis(
+      timestampDiff(
+          TimeIntervalUnit.MONTH,
+          "2018-07-03 11:11:11".toTimestamp,
+          "2018-09-05 11:11:11".toTimestamp),
+      "timestampDiff(MONTH, " +
+        "'2018-07-03 11:11:11'.toTimestamp, '2018-09-05 11:11:11'.toTimestamp)",
+      "TIMESTAMPDIFF(MONTH, " +
+        "TIMESTAMP '2018-07-03 11:11:11', TIMESTAMP '2018-09-05 11:11:11')",
+      "2")
+
+    testAllApis(
+      timestampDiff(TimeIntervalUnit.MONTH, "2016-06-15".toDate, "2016-04-18".toDate),
+      "TIMESTAMPDIFF(MONTH, '2016-06-15'.toDate, '2016-04-18'.toDate)",
+      "TIMESTAMPDIFF(MONTH, DATE '2016-06-15', DATE '2016-04-18')",
+      "-1")
+
+    testAllApis(
+      timestampDiff(TimeIntervalUnit.MONTH, "2016-06-15".toDate, "2016-04-14".toDate),
```

Comment:
We should test the full matrix. MONTH on date, timestamp is missing.

Maybe we can make the tests more generic similar to timestamp add. We should test every combination of unit, type, type.